extensions/amp-cxense-player/0.1/amp-cxense-player.js

```diff
@@ -0,0 +1,319 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {CSS} from '../../../build/amp-cxense-player-0.1.css';
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {loadPromise} from '../../../src/event-helper';
+import {appendParamStringToUrl, parseUrl} from '../../../src/url';
+import {dashToCamelCase} from '../../../src/string';
+import {setStyles} from '../../../src/style';
+const extend = require('extend');
+
+const cxDefaults = {
+  apiHost: 'https://api.widget.cx',
+  embedHost: 'https://embed.widget.cx',
+  distEmbedApp: '/app/player/m4/dist/',
+  debugEmbedApp: '/app/player/m4/debug/',
+  attrs: {
+        // can't access the window.top.location.href from the iframe.
+    share: {
+      enable: false,
+    },
+  },
+};
+
+class AmpCxense extends AMP.BaseElement {
+
+    /** @override */
+    preconnectCallback(onLayout) {
+      this.preconnect.url(cxDefaults.apiHost, onLayout);
+      this.preconnect.prefetch(
+            cxDefaults.embedHost + cxDefaults.distEmbedApp, 'iframe'
+        );
+    }
+
+    /** @override */
+    isLayoutSupported(layout) {
+      return isLayoutSizeDefined(layout);
+    }
+
+    /** @override */
+    buildCallback() {
+      this.element.classList.add('amp-cxense-player');
+
+      if (!this.getPlaceholder()) {
+        this.buildWidgetPlaceholder_();
+      }
+    }
+
+    /** @override */
+    layoutCallback() {
+      const self = this;
+
+        /** @private @const {Element} */
+      this.iframe_ = this.element.ownerDocument.createElement('iframe');
+
+      this.iframe_.setAttribute('frameborder', '0');
+      this.iframe_.setAttribute('allowfullscreen', 'true');
+      this.iframe_.width = this.element.getAttribute('width');
+      this.iframe_.height = this.element.getAttribute('height');
+      this.iframe_.src = this.getIframeSrc_();
+      setStyles(this.iframe_, {display: 'none'});
+
+      this.element.appendChild(this.iframe_);
+      this.listenForPostMessages_();
+
+      return loadPromise(this.iframe_).then(ret => {
+        if (self.placeholder_) {
+          setStyles(this.placeholder_, {display: 'none'});
+        }
+        setStyles(self.iframe_, {display: ''});
+        self.applyFillContent(self.iframe_);
+        return ret;
+      });
+    }
+
+    /** @override */
+    pauseCallback() {
+      this.postMessage_({type: 'mpf.video.pause'});
+      return true;
+    }
+
+    /** @override */
+    unlayoutCallback() {
+      this.iframe_.setAttribute('src', 'about:blank');
+      this.iframe_.parentNode.removeChild(this.iframe_);
+      return true;
+    }
+
+    /** @private */
+    listenForPostMessages_() {
+      const hostRegExp = new RegExp(
+            '^' + parseUrl(cxDefaults.embedHost).hostname
+        );
+
+      this.getWin().addEventListener('message', e => {
+        if (e && e.data) {
```

Comment:
First filter by `e.source == this.iframe.contentWindow` please and return if the message is from a different window.